### DISCUSSION TITLE
Classes for Patches and Patchbanks

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#179] [Show and tell]
This is the method that I've pretty much standardized for storing patches with associated information and banks of patches.

I use these for librarians, for example.

```
-- Intialize Classes
defaultSysex = {0xf0,0x00,0x00,0x2f,0x03,0x02,0x00,0x00,0x00,0x00,0x00,0x09,0x00,0x0c,0x00,0x00,0x00,0x00,0x0b,0x02,0x06,0x02,0x0f,0x0f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x0f,0x0f,0x00,0x00,0x09,0x0c,0x00,0x00,0x00,0x00,0x08,0x08,0x00,0x00,0x02,0x0c,0x01,0x07,0x0f,0x07,0xf7} -- This particular patch is for a Jupiter 8
defaultSysexMB = MemoryBlock()
defaultSysexMB:createFromTable(defaultSysex)
patch = {name = "No Name", author = "Default Author", category = "None", sysex = defaultSysexMB} -- Initialize table structure. Category only allows a single value currently.  Maybe redefine as a table. This is difficult to implement. In practice pulling a value from a table, of a table, of a table, doesn't work.

function patch:new (o)
    o = o or {} -- create object if user does not provide one
    setmetatable(o, self) -- define prototype
    self.__index = self
    return o
end

patchBank = {} -- each index is the patch ID. As we add patches we should call them as patchBank[1] = patch:new(). We can then access data by patchBank[1].name etc.
function patchBank:new(o)
    o = o or {} -- create object if user does not provide one
    setmetatable(o,self) -- define prototype
    self.__index = self
    return o
end

synthBank = patchBank:new() -- create the Synth Bank.
```

Elements within the patchbank can be accessed and copied like this

`synthBank[i] = synthBank2[j]`

You have full access to elements like

```
tempName = synthBank[i].name
tempAuthor = synthBank[i].author
tempCategory = synthBank[i].category
tempSysex = synthBank[i].sysex
```

I strip sysex headers off during import and only store the data segment of the sysex.  I rebuild the header when I send it to the synth. The reason for this is because the header normally contains the destination bank and program number, and we want that to be dynamically assigned.

**Comment by dobo365** [comment]
Interesting.
I'm keeping everything as MemoryBlock and handle all with getByte(), getRange() including name, category...
Banks are also memoryblocks.
Easy to save/read to/from file as well.
I keep the headers because they are used to recognize the synth and manufacturer.

**Reply by Godlike-Productions** [reply]
If a synth has categories or names within the dump then I'll keep them there, but the Jupiter 8, I'm doing all of that within the plugin itself, so strings works fine, and they need to be converted to strings to display in labels and list boxes anyway.

You can define any of the elements as memoryblocks easily, just make the declaration as I did with Sysex.

**Reply by dnaldoog** [reply]
Can you explain where <code>synthBank[i] = synthBank2[j]</code> comes from?

**Reply by Godlike-Productions** [reply]
That's just an example of how you can copy the record from one Bank to another.

You can also do it at a lower level like this

```synthBank[4].sysex = synthBank2[2].sysex```

synthBank and synthBank 2 would have to have already been defined and populated though using

```synthBank = patchBank:new() -- create the Synth Bank.```

**Reply by dnaldoog** [reply]
> That's just an example of how you can copy the record from one Bank to another.
> 
> You can also do it at a lower level like this
> 
> `synthBank[4].sysex = synthBank2[2].sysex`
> 
> synthBank and synthBank 2 would have to have already been defined and populated though using
> 
> `synthBank = patchBank:new() -- create the Synth Bank.`

it looks like you have a table of synthBank[n] somewhere, but I don't see it declared or explained where it comes from? Not to worry - I think I can get where you are coming from with that!

Also with this code  ....

<pre>
synthBank = patchBank:new() -- create the Synth Bank.

You have full access to elements like

tempName = synthBank[i].name
tempAuthor = synthBank[i].author
tempCategory = synthBank[i].category
tempSysex = synthBank[i].sysex
</pre>

then how do you get access to synthBank[i].__name__ when __name__ is a key of table __patch__ which doesn't belong to table/class  patchBank?

 I can't see the connection, unless there is some inheritance going on that I am missing? 

I don't see what functions the class declarations are serving as there are no member functions in the class??

**Reply by dnaldoog** [reply]
So I removed the reference to array indices in the table synthBank[i] but this code doesn't access that table patch. Nothing is working -  I think it needs to be a member of patchBank{}

<pre>
myMethod = function(mod, value, source)
    -- Intialize Classes
   defaultSysex = {0xf0,0x00,0x00,0x2f,0x03,0x02,0x00,0x00,0x00,0x00,0x00,0x09,0x00,0x0c,0x00,0x00,0x00,0x00,0x0b,0x02,0x06,0x02,0x0f,0x0f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x0f,0x0f,0x00,0x00,0x09,0x0c,0x00,0x00,0x00,0x00,0x08,0x08,0x00,0x00,0x02,0x0c,0x01,0x07,0x0f,0x07,0xf7} -- This particular patch is for a Jupiter 8
    defaultSysexMB = MemoryBlock()
    defaultSysexMB:createFromTable(defaultSysex)
    patch = {name = "No Name", author = "Default Author", category = "None", sysex = defaultSysexMB} -- Initialize table structure. Category only allows a single value currently.  Maybe redefine as a table. This is difficult to implement. In practice pulling a value from a table, of a table, of a table, doesn't work.

    function patch:new(o)
        o = o or {} -- create object if user does not provide one
        setmetatable(o, self) -- define prototype
        self.__index = self
        return o
    end

    patchBank = {} -- each index is the patch ID. As we add patches we should call them as patchBank[1] = patch:new(). We can then access data by patchBank[1].name etc.
    function patchBank:new(o)
        o = o or {} -- create object if user does not provide one
        setmetatable(o, self) -- define prototype
        self.__index = self
        return o
    end

    synthBank = patchBank:new() -- create the Synth Bank.d

    tempName = synthBank.name or "nil"
    tempAuthor = synthBank.author or "nil"
    tempCategory = synthBank.category or "nil"
    tempSysex = "nil" --synthBank.sysex or "nil" below should be tempSysex:toHexString(1)

    console(
        string.format(
            "tempName %s\ntempAuthor %s\ntempCategory %s\ntempSysex %s",
            tempName,
            tempAuthor,
            tempCategory,
            tempSysex
        )
    )
end




</pre>

I still don't know quite what you are trying to achieve but it is interesting :)


Are you just trying to bundle related data to the actual sysex message and populate a panel depending on what is selected?

**Reply by dnaldoog** [reply]
I think you would need to do something like this. __This works!!__

<pre>
myMethod=function(mod,value,source)

-- Intialize Classes
defaultSysex = {0xf0,0x00,0x00,0x2f,0x03,0x02,0x00,0x00,0x00,0x00,0x00,0x09,0x00,0x0c,0x00,0x00,0x00,0x00,0x0b,0x02,0x06,0x02,0x0f,0x0f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x0f,0x0f,0x00,0x00,0x09,0x0c,0x00,0x00,0x00,0x00,0x08,0x08,0x00,0x00,0x02,0x0c,0x01,0x07,0x0f,0x07,0xf7} -- This particular patch is for a Jupiter 8
defaultSysexMB = MemoryBlock()
defaultSysexMB:createFromTable(defaultSysex)
--patch = {name = "No Name", author = "Default Author", category = "None", sysex = defaultSysexMB} -- Initialize table structure. Category only allows a single value currently.  Maybe redefine as a table. This is difficult to implement. In practice pulling a value from a table, of a table, of a table, doesn't work.

--function patch:new (o) -- DON'T NEED THIS AS I SEE IT
 --   o = o or {} -- create object if user does not provide one
 --   setmetatable(o, self) -- define prototype
 --   self.__index = self
 --   return o
--end


patchBank = {} -- each index is the patch ID. As we add patches we should call them as patchBank[1] = patch:new(). We can then access data by patchBank[1].name etc.
function patchBank:new(o)
    o = o or {} -- create object if user does not provide one
    setmetatable(o,self) -- define prototype
    self.__index = self
	self.patch = {name = "No Name", author = "Default Author", category = "None", sysex = defaultSysexMB} -- Initialize table structure. Category only allows a single value currently.  Maybe redefine as a table. This is difficult to implement. In practice pulling a value from a table, of a table, of a table, doesn't work.
    return o
end

function patchBank:getName()
return self.patch.name
end

function patchBank:getSysex()
return self.patch.sysex
end

function patchBank:setName(myNewName)
self.patch.name=myNewName
end

local synthBank = patchBank:new() -- create the Synth Bank.d
synthBank:setName("I am nobody")

tempName = synthBank:getName() or "nil"
tempAuthor = synthBank.author or "nil" -- no getter or setter methods implemented
tempCategory = synthBank.category or "nil" -- no getter or setter methods implemented
tempSysex = synthBank:getSysex() or "nil" 

console(string.format("tempName %s\ntempAuthor %s\ntempCategory %s\ntempSysex %s",tempName,tempAuthor,tempCategory,tempSysex:toHexString(1)))


end
</pre>
__ this outputs ...__

<pre>
LUA>> tempName I am nobody
tempAuthor nil
tempCategory nil
tempSysex f0 00 00 2f 03 02 00 00 00 00 00 09 00 0c 00 00 00 00 0b 02 06 02 0f 0f 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0f 0f 00 00 09 0c 00 00 00 00 08 08 00 00 02 0c 01 07 0f 07 f7
</pre>

**Reply by Godlike-Productions** [reply]
Oh sorry, I forgot one line.  To put a patch into patchbank use

```patchBank[1] = patch:new()```

You can also go patchBank[q] = patch:new({name = "whatever name", author = "something here"}) to overwrite the prototype.

If you don't supply name, author, category or sysex, it will use whatever you define in the prototype.

Basically patchbank is a class containing patches.
Patch is a class with name, category, author and sysex.

The purpose of doing this is to keep the large structures organized, so patchBank[i] automatically refers to all of the associated names, authors etc, rather than having to maintain multiple tables.

It also makes moving patches between banks trivial.

Below is the screengrab of the librarian I've developed.  Jupiter column is on the actual synth (it tracks patch names in the plugin). I will export sysex from clipboard, have a library backup, and can import into the Import column.  
![Jupiter 8 Morph 51 Alpha Librarian](https://user-images.githubusercontent.com/13851825/108622775-f2e4ab00-7475-11eb-8e68-8c7b90b3a77e.png)

**Reply by dnaldoog** [reply]
Oh! Now I see - I should have seen that. It would be interesting to set up [inheritance](http://lua-users.org/wiki/InheritanceTutorial) for that. I have tried but it does my head in - also it's usually not necessary for me, but well done and a nice example!

**Reply by Godlike-Productions** [reply]
There is a chapter on inheritance in the LUA manual. It fully supports it.

**Reply by dnaldoog** [reply]
<pre>
myMethod = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)
-- Intialize Classes
defaultSysex = {0xf0,0x00,0x00,0x2f,0x03,0x02,0x00,0x00,0x00,0x00,0x00,0x09,0x00,0x0c,0x00,0x00,0x00,0x00,0x0b,0x02,0x06,0x02,0x0f,0x0f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x0f,0x0f,0x00,0x00,0x09,0x0c,0x00,0x00,0x00,0x00,0x08,0x08,0x00,0x00,0x02,0x0c,0x01,0x07,0x0f,0x07,0xf7} -- This particular patch is for a Jupiter 8
defaultSysexMB = MemoryBlock()
defaultSysexMB:createFromTable(defaultSysex)
patch = {name = "No Name", author = "Default Author", category = "None", sysex = defaultSysexMB} -- Initialize table structure. Category only allows a single value currently.  Maybe redefine as a table. This is difficult to implement. In practice pulling a value from a table, of a table, of a table, doesn't work.

function patch:new (o)
    o = o or {} -- create object if user does not provide one
    setmetatable(o, self) -- define prototype
    self.__index = self
    return o
end

patchBank = {} -- each index is the patch ID. As we add patches we should call them as patchBank[1] = patch:new(). We can then access data by patchBank[1].name etc.
function patchBank:new(o)
    o = o or {} -- create object if user does not provide one
    setmetatable(o,self) -- define prototype
    self.__index = self
    return o
end
patchBank =patch:new()
-- patchBank = patch --patch:new() or this!
synthBank = patchBank:new() -- create the Synth Bank.
--Elements within the patchbank can be accessed and copied like this

--synthBank  = synthBank2[j]

--You have full access to elements like

tempName = synthBank.name
tempAuthor = synthBank.author
tempCategory = synthBank.category
tempSysex = synthBank.sysex


tempName = synthBank.name or "nil"
tempAuthor = synthBank.author or "nil"
tempCategory = synthBank.category or "nil"
tempSysex = synthBank.sysex or "nil" 

console(string.format("tempName %s\ntempAuthor %s\ntempCategory %s\ntempSysex %s",tempName,tempAuthor,tempCategory,tempSysex:toHexString(1)))

end
</pre>

Now I have got your code working, but still I think the declaration of _patch:new()_ is unnecessary. I can equally just use _patch{}_ directly as a static initializer to patchBank as you are doing through patch:new(), but that's not a problem. 

In my example I only use one class object and the _patch{}_ table is a member of that class, which is perhaps easier to understand. I'm no expert on OOP - Anyway, very nice example as I said.

**Reply by Godlike-Productions** [reply]
Oh nice, thanks for the example. I'll look at this method for the next plugin.  Too far gone in this one at the moment.

**Reply by Godlike-Productions** [reply]
Just coming back to this. 

patch:new() is important in many cases because it ensures that a new patch is valid as a fallback., I'll experiement with patch{}, it's probably equally valid.  I've discovered that if you have nil in any part of the structure, then it will fail if you attempt to export json, and can potentially crash ctrlr.  

I'm concerned about this

```
tempName = synthBank.name or "nil"
tempAuthor = synthBank.author or "nil"
tempCategory = synthBank.category or "nil"
tempSysex = synthBank.sysex or "nil" 
```

What's the purpose of this?  Does this replace nil with the string "nil" if something is empty?  If that's the case, that's quite an elegant solution to the issue.  Just trying to understand.

**Comment by dnaldoog** [comment]
I'm guessing that was what I was intending, because _json_ doesn't accept type **nil**, so you would be writing a String value to the file. When you read it back in you could then interpret "nil" how you wanted. I guess you could also do a _boolean_ and check for truth when you load the _json_ file back into your program: 🤔

<pre>
tempName = synthBank.name or false
tempAuthor = synthBank.author or false
tempCategory = synthBank.category or false
tempSysex = synthBank.sysex or false 
</pre>

`if tempCategory==false then ...` for example or

`if type(tempCategory) ~= "string"` then ...
